text/0000-router-extensibility.md

```diff
@@ -0,0 +1,165 @@
+- Start Date: 2018-01-13
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# Router DSL extensibility
+
+## Summary
+
+Adding a public hook to the Ember Router and one to Ember.Route would allow for addon authors and advanced app developers to enrich the route definition process in expressive ways that are not currently possible with existing hooks.
+
+## Motivation
+
+This RFC proposes 2 changes to support addon authors and app developers in enhancing the expressiveness of the route map. First, a hook would be added to allow the DSL implementation to be specified or extended. Second, any route options specified in the map function would be provided to the route. These changes would work hand-in-hand, with the DSL specifying options that would be available to the route.
+
+### Example use cases
+
+An authentication addon could allow a developer to wrap one or more routes with `this.authenticated` to express that certain route hierarchies require authentication before entering:
+
+```js
+import Router from 'my-auth-addon/router';
+
+Router.map(function(){
+  this.route('about');
+  this.route('login');
+  this.authenticated(function(){
+    this.route('billing');
+    this.route('preferences');
+  });
+});
+
+export default Router;
+```
+
+The router map for a mobile app that uses a "navigation stack" UI could indicate that certain routes should slide up a new layer:
+
+```js
+import RouterDSL from '@ember/routing/dsl';
+
+let Router = EmberRouter.extend({
+  dslFactory: RouterDSL.extend({
+    layer(routeName, opts = {}, callback=null) {
+      let optionsWithLayer = Object.assign({ newLayer: true }, opts);
+      return this.route(routeName, optionsWithLayer, callback);
+    };
+  });
+});
+
+Router.map(function(){
+  this.route('agenda', function() {
+      this.route('schedule-item', { '/schedule-items/:schedule_item_id' });
+      this.layer('my-schedule', function() {
+        this.route('schedule-item', { '/schedule-items/:schedule_item_id' });
+      });
+    });
+  });
+})
+```
+
+In both examples, a route base class or mixin could be used to interpret options provided to the route and customize its behavior accordingly.
+
+### Background
+
+The router.js file contains the high-level definition of the app's routes, and is often the first place a developer looks to begin becoming familiar with a new app or to orient herself when beginning a new feature. It currently allows the expression of the following aspects of the app's routes in the function passed to `Router#map`:
```

Comment:
I'm unfamiliar with some of the nuances of RFCs, but this paragraphs feels better suited for the motivation section. Or at least the concepts communicated here.